File: P1.py
# next 
# TC : O(n)
# SC : O(h)

# hasNext
# TC : O(1)
# SC : O(1)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:
    def __init__(self, root: Optional[TreeNode]):
        self.stack = []
        self.left_most_inorder(root)
        # print("printing stack everytime", (self.stack))
    
    def left_most_inorder(self, root):
        while root:
            self.stack.append(root)
            # print(self.stack,"             ok             ")
            root = root.left

    def next(self) -> int:
        top_node = self.stack.pop()
        # print("printing top most node", top_node, "          ", self.stack)
        if top_node.right:
            self.left_most_inorder(top_node.right)
        return top_node.val
        

    def hasNext(self) -> bool:
        return len(self.stack) > 0


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: p1.java
// Time Complexity : hasNext - O(1)
// next - O(1) amortized
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Yes. Could not figure out the code. Had to refer lecture




/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {
    //Stack to keep track of which nodes are pending to be processed
    Stack<TreeNode> stk;
    public BSTIterator(TreeNode root) {
        stk = new Stack<>();
        //Getting the leftmost element and storing other elements which came before the leftmost in the stack
        while(root!=null){
            stk.push(root);
            root = root.left;
        }
    }
    
    public int next() {
        //get the element on top of the stack and process it. Run DFS on the right of the current element 
        TreeNode root = stk.pop();
        TreeNode head = root.right;

        while(head!=null){
            stk.push(head);
            head = head.left;
        }

        return root.val;
    }
    
    public boolean hasNext() {
        //Check if stack is empty or not
       return !stk.isEmpty();
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */